setup.go

```diff
@@ -0,0 +1,132 @@
+package influxdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+)
+
+// Setup sets up a new influxdb server.
+// It requires a client be set up with a username and password.
+// If successful will add a token to the client.
+// RetentionPeriodHrs of zero will result in infinite retention.
+func (c *Client) Setup(ctx context.Context, bucket, org string, RetentionPeriodHrs int) (*SetupResult, error) {
+	if c.username == "" || c.password == "" {
+		return nil, errors.New("a username and password is requred for a setup")
+	}
+
+	inputData, err := json.Marshal(SetupRequest{
+		Username:           c.username,
+		Password:           c.password,
+		Org:                org,
+		Bucket:             bucket,
+		RetentionPeriodHrs: RetentionPeriodHrs,
+	})
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("POST", c.url.String()+"/setup", bytes.NewBuffer(inputData))
+	if err != nil {
+		return nil, err
+	}
+	req = req.WithContext(ctx)
+	req.Header.Add("Content-Type", "application/json; charset=utf-8")
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	setupResult := &SetupResult{}
+	if err := json.NewDecoder(resp.Body).Decode(setupResult); err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	if setupResult.Code != "conflict" && resp.StatusCode == http.StatusCreated && setupResult.Auth.Token != "" {
+		c.l.Lock()
```

Comment:
I'm not sure I follow why the lock is required here? Is this necessary?

Reply:
To prevent a "harmless" data race.